### PR TITLE
net: sockets: Verify that TCP context exists when updating window length

### PR DESCRIPTION
**Pull request by rlubos**
Socket may still consume received data after TCP connection was closed and TCP context deallocated, therefore it is needed to verify that TCP context still exists before using it. Otherwise TCP layer throws an error, which can be seen for example in `http_get` sample.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

### CODE REVIEW

**Comment by pfalcon** [comment]
> Otherwise TCP layer throws an error, which can be seen for example in http_get sample.

Which error? Do we have a ticket open for it? Can the error be quoted in the commit message for clarity?

**Comment by rlubos** [comment]
image

**Comment by pfalcon** [comment]
This is not a reliable solution. 

> Socket may still consume received data after TCP connection was closed and TCP context deallocated

A TCP context *is* a socket. In POSIX semantics, a socket can't be deallocated behind the owning application's back. And if it's deallocated, that slot can be reused, and the code will access completely different context.

And that's one of the big issues in the Zephyr IP stack that it **can** deallocate net context behind its owner's back. And as usual, it's a contention point whether to resolve it on the level of the IP stack, or on the level of BSD socket shim.

Summing up, I'd prefer to not apply this patch - it masks issue rather than resolves it. Instead, should consider ways to resolve the underlying issue described above.

**Comment by rlubos** [comment]
While I totally agree that deallocation of TCP context **is** a problem, and believe me, I was strongly suprised when I figured out what is going on, I do not see anything wrong with providing means to avoid the erroneous situation on a higher layer. Yes, it may mask the problem, but it prevents potential users from hitting the error until proper TCP context handling is introduced, which will happend I don't know when. And hitting errors in staple examples is a problem IMHO - it discourages people to use  Zephyr.

Of course I don't deny that the problem exists and appropritate ticket should be created, so it is not overlooked during the transition to sockets.

**Comment by pfalcon** [comment]
> I do not see anything wrong with providing means to avoid the erroneous situation on a higher layer.

My concern with changes like this is that they're workarounds adding extra code. When the root issue is fixed, nobody will remember to remove it. Over time, we'll accumulate a bunch of cruft which will show up in size comparisons with other IP stacks, and that will also make Zephyr less tempting for people.

> but it prevents potential users from hitting the error until proper TCP context handling is introduced

That's why I don't give it -1, just -0. I'm sure other folks will approve it. The alternative is to withdraw this patch now, and pull together for the underlying fix. With more than 1 people seeing the issue, we hopefully will be able to cover it in a couple of weeks. Or if you say that you think that it's important that this workaround-style fix gets into the 1.12 release, I can +1 it either.

**Comment by pfalcon** [comment]
> appropritate ticket should be created

net_context lifetime issues are tracked as https://github.com/zephyrproject-rtos/zephyr/issues/8008

**Comment by rlubos** [comment]
Closing this one as I'm not able to reproduce this anymore.

**Comment by pfalcon** [comment]
@rlubos : Fixing this in the more general way (#8008) is still on my todo.
